src/components/connectAdvanced.js

```diff
@@ -124,22 +99,52 @@ export default function connectAdvanced(
 
         this.version = version
         this.renderCount = 0
-        this.store = props[storeKey] || context[storeKey]
+        const store = props[storeKey] || context[storeKey]
         this.propsMode = Boolean(props[storeKey])
         this.setWrappedInstance = this.setWrappedInstance.bind(this)
 
-        invariant(this.store,
+        invariant(store,
           `Could not find "${storeKey}" in either the context or props of ` +
           `"${displayName}". Either wrap the root component in a <Provider>, ` +
           `or explicitly pass "${storeKey}" as a prop to "${displayName}".`
         )
+        const storeState = store.getState()
 
+        const childPropsSelector = this.createChildSelector(store)
+        this.state = {
+          props,
+          childPropsSelector,
+          childProps: {},
+          store,
+          storeState,
+          error: null,
+        }
         this.state = {
-          updater: this.createUpdater()
+          ...this.state,
+          ...Connect.getChildPropsState(props, this.state)
         }
         this.initSubscription()
       }
 
+      static getChildPropsState(props, state) {
+        try {
+          const nextProps = state.childPropsSelector(state.store.getState(), props)
+          if (nextProps === state.childProps) return null
+          return { childProps: nextProps }
+        } catch (error) {
+          return { error }
+        }
+      }
+
+      static getDerivedStateFromProps(props, state) {
+        if (connectOptions.pure && shallowEqual(props, state.props) || state.error) return null
```

Comment:
Maybe add parentheses around the `pure && shallowEqual()` bit just to clarify the comparison?